### PR TITLE
Support for getting / injecting shared Maven components in JibMavenPluginExtension

### PR DESCRIPTION
**Pull request by stefanocke**
Pull request for #3036.

### CODE REVIEW

**Comment by stefanocke** [comment]
@googlebot I signed it!

**Comment by stefanocke** [comment]
I don't know how to increase test coverage further. The remaining 4 uncovered lines are all within the Mojos. AFAIU, the Mojos are covered by the integration tests. However, they seem not to be part of coverage measurements...
I don't think I should write some new kind of Mojo tests now. So I leave it like it is. Please tell me what to do as part of your review.

**Review by chanseokoh — COMMENTED** [review]
This "Maven & JSR-330" stuff (Sisu?) looks like a magic.

**Comment by chanseokoh** [comment]
@stefanocke question: I see the change is to inject Jib extensions, not that it's to inject a Maven component in an Jib extension. Just in case, I wonder if Sisu can't just inject Maven components for Jib extensions loaded by the JDK server loader? (If that's indeed the case, we wouldn't need this change?)

File: jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/JibPluginConfiguration.java [hunk 1]
```diff
@@ -346,6 +349,12 @@ public String getExtensionClass() {
 
   @Component protected SettingsDecrypter settingsDecrypter;
 
+  @Inject private Set<JibMavenPluginExtension<?>> injectedPluginExtensions = Collections.emptySet();
+
+  protected Set<JibMavenPluginExtension<?>> getInjectedPluginExtensions() {
+    return injectedPluginExtensions;
+  }
+
```
[the remark this answers is not in the document]
Comment (chanseokoh) [reply jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/JibPluginConfiguration.java:357]: On second thought, it's okay to have this class know `JibMavenPluginExtension` and have instances injected this way. However, I still think this class shouldn't have a reference to `MavenProjectProperties`. It creates a cyclic dependency graph; `MavenProjectProperties` is a higher-level abstraction that should sit above `JibPluginConfiguration`.

File: jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/JibPluginConfiguration.java [hunk 2]
```diff
@@ -821,8 +830,24 @@ void setSession(MavenSession session) {
...
+  void setDescriptor(PluginDescriptor descriptor) {
+    this.descriptor = descriptor;
+  }
+
   @Nullable
   String getProperty(String propertyName) {
     return MavenProjectProperties.getProperty(propertyName, project, session);
   }
+
+  protected MavenProjectProperties getMavenProjectProperties(
```
[the remark this answers is not in the document]
Comment (stefanocke) [reply jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/JibPluginConfiguration.java:843]: @chanseokoh to be honest, the only reason I moved this here was to achieve the test coverage goal for my commits. Initially I added the line getInjectedPluginExtensions() in every Mojo.
Unfortunately, the Mojos seems only to be tested in integration tests, which seem not to count for  the test coverage goal. (I don't know exactly. )
So basically, by moving it up to JibMavenPluginExtension I could write a basic test for it. and my ratio between changed code lines and covered lines improved sufficiently.
I dont't know how else I can achieve test coverage goal. I need your advise here...

File: jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/JibPluginConfiguration.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (stefanocke) [reply jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/JibPluginConfiguration.java:357]: Ok, thank you. Glad to here, `JibMavenPluginExtension` can stay here. For `MavenProjectProperties` please see below.

**Comment by stefanocke** [comment]
@chanseokoh , the short answer is: no.

The longer answer:
The goal was to inject a Maven component in the jib extension, which I did in the other pull request (the MavenDependencyResolver). 
But to achieve this, the extension must be instantiated and processed by the Sisu Container. That's why I added some annotations to the extension and some META-INF/sisu.  
If we just load the extension by service loader, it won't be the one that has been created and processed by Sisu. I will just be a one that the Service Loader would create by calling "newInstance()".
Of course we could try to "post-process" that instance. But then we have to problems to solve:

- Somehow to call Sisu Dependency injection programmatically, which is probably not possible at all, or at least not by Maven public APIs.
- And we would still have to do this in jib maven plugin. With probably even much more code than now.

Or do you mean the extension could do the dependency injection by itself, for exampe in its constructor?
Then - again - we would need some public API in Maven again to to this. Or simply something like "getSharedComponent()".
But this was exactly the "old" approach... the deprecated methods in Maven session.

So to summarize: I am convinced the approach I have choosen is the right one. It is compliant to the only existing Maven documentation regarding Sisu / DI. And every other approch would mean more code and higher probability to break.

**Comment by chanseokoh** [comment]
> the short answer is: no.

Yeah, I expected this. Didn't really thought Sisu can inject things into an instance created by the service loader. Was wondering just in case. I think this approach looks good, and perhaps this is what we can recommend going forward.

File: jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/JibPluginConfiguration.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (chanseokoh) [reply jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/JibPluginConfiguration.java:843]: Forget about the codecov coverage. It's okay if the check for the PR fails. I think it's right that `MavenProjectProperties` should not appear in this class.

[the remark this answers is not in the document]
Comment (stefanocke) [reply jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/JibPluginConfiguration.java]: Ok, I reverted the regarding refactoring commit.

File: jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/JibPluginConfigurationTest.java [hunk 3]
```diff
@@ -342,4 +350,12 @@ public void testIsContainerizable_directory() {
     projectProperties.setProperty("jib.containerize", "project2");
     Assert.assertFalse(testPluginConfiguration.isContainerizable());
   }
+
+  @Test
+  public void testGetMavenProjectProperties() {
+    TempDirectoryProvider tdp = new TempDirectoryProvider();
```
[the remark this answers is not in the document]
Comment (stefanocke) [reply jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/JibPluginConfigurationTest.java]: Obsolete. That was just my quite artificial test for the getMavenProjectProperties() refactoring.

File: jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/MavenProjectPropertiesExtensionTest.java [hunk 4]
```diff
@@ -450,4 +451,76 @@ public void testRunPluginExtensions_runtimeExceptionFromExtension() {
...
+            (buildPlan, properties, extraConfig, mavenData, logger) -> {
+              logger.log(LogLevel.ERROR, "awesome error from my extension");
+              return buildPlan.toBuilder().setUser("user from injected extension").build();
+            });
+
+    FooExtension loadedExtension =
+        new FooExtension(
+            (buildPlan, properties, extraConfig, mavenData, logger) -> {
+              return buildPlan.toBuilder().setUser("user from extension").build();
+            });
```
[the remark this answers is not in the document]
Comment (stefanocke) [reply jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/MavenProjectPropertiesExtensionTest.java]: ok

**Review by chanseokoh — APPROVED** [review]
Thank you for your contribution!

**Comment by chanseokoh** [comment]
Thanks again for this cool feature!
